teps/0122-reproducibility-of-complete-build-instructions.md

```diff
@@ -0,0 +1,279 @@
+      annotations: null
+    - entryPoint: ""
+      arguments:
+        - $(params.EXTRA_ARGS)
+        - --dockerfile=$(params.DOCKERFILE)
+        - --context=$(workspaces.source.path)/$(params.CONTEXT)
+        - --destination=$(params.IMAGE)
+        - --digest-file=$(results.IMAGE_DIGEST.path)
+      environment:
+        container: build-and-push
+        image: docker-pullable://gcr.io/kaniko-project/executor@sha256:68bb272f681f691254acfbdcef00962f22efe2f0c1e287e6a837b0abe07fb94b
+      annotations: null
+    - entryPoint: |
+        set -e
+        echo $(params.IMAGE) | tee $(results.IMAGE_URL.path)
+      arguments: null
+      environment:
+        container: write-url
+        image: docker-pullable://bash@sha256:fc742d0c3d9d8f5fb2681062398c04b710cd08c46dac1a8f0a5515687018acb9
+      annotations: null
+```
+
+## Requirements
+- It must be possible to use the provenance generated by Tekton Chains for an artifact to create a bit by bit identical (best effort) artifact.
+  - Much of this will depend on the Task itself; note that the SLSA L4 required is "best effort".
+  - The provenance should contain all of the Task definition and runtime information required to reproduce the build.
+- The shas of the images run should continue to be included regardless (as this is part of the dependencies / build instructions)
+- It must be possible for users to construct policies based on the provenance that would allow them to determine if the build is ok - for example if they have policies around what pipeline tasks are acceptable to use, what parameters are acceptable, whether or not sidecars are allowed etc.
+  - It is acceptable that a policy might need to fetch the source tasks and pipelines from version control (or wherever they are stored) but regardless it must be possible for a policy engine to make the decisions it needs from the provenance (and values referenced by the provenance)
+
+## Proposal
+In order to reproduce the build, we need to be able to extract enough information from the provenance to recreate the task run or pipeline run under the same set of configuration as was used when creating the build. The list below shows which fields are required in the provenance for re-creating the task run.
+- [Task Specification](#task-specification)
+- [TaskRef](#taskref)
+- [TaskRun Specification](#taskrun-specification)
+- [Step Specification](#step-specification)
+- [Configuration feature flags](#configuration-feature-flags)
+- [Reproducibility Reason Field if not reproducible](#reproducibility-reason-field-if-not-reproducible) 
+- [Tekton Pipelines and Chains Version](#tekton-pipelines-and-chains-version)
+
+This will allow us to create a provenance for a task run as [shown](#provenance-for-executed-taskrun).
+
+**Generating provenance from the pipeline run is still a [WIP](https://github.com/tektoncd/chains/pull/436). Therefore, this proposal only covers reproducibility of task run.** 
```

Comment:
FWIW, this is about to be merged.

Reply:
Oh that's great!

Reply:
It's merged :sunglasses:

Reply:
That's wonderful. I will update the TEP in that case.